### PR TITLE
Associations: 1

### PR DESCRIPTION
**Pull request by akshay1y2**

### CODE REVIEW

**Comment by akshay1y2** [comment]
Tasks: https://docs.google.com/document/d/17GD0udHhXzTmRNOZZy_nsysIBnafIomr_xi4OkL-jYQ/edit?pli=1#

File: app/controllers/orders_controller.rb [hunk 1]
```diff
@@ -29,6 +29,7 @@ def edit
   # POST /orders.json
   def create
     @order = Order.new(order_params)
+    @order.user = User.find_by_id(session[:user_id])
```
Comment (siddharth28) [inline app/controllers/orders_controller.rb]: @akshay1y2 please follow DRY.
`User.find_by_id(session[:user_id])` repeating code.

File: app/controllers/orders_controller.rb [hunk 2]
```diff
@@ -69,6 +70,14 @@ def destroy
     end
   end
 
+  def user_orders
+    @orders = User.find_by_id(session[:user_id]).orders.map do |order|
+      { order: order,
+        amount: order.line_items.inject(0) { |price, item| price += item.product.price },
```
Comment (siddharth28) [inline app/controllers/orders_controller.rb]: @akshay1y2 `n+1` queries.

File: app/models/cart.rb [hunk 3]
```diff
@@ -1,5 +1,7 @@
 class Cart < ApplicationRecord
   has_many :line_items, dependent: :destroy
+  has_many :products, through: :line_items
+  has_many :enabled_products, ->{ where enabled: true }, through: :line_items, source: :product
```
Comment (siddharth28) [inline app/models/cart.rb]: @akshay1y2 why are we not creating a scope for this?
I have already asked you twice for this very thing.

File: app/models/product.rb [hunk 4]
```diff
@@ -1,16 +1,22 @@
...
 
   validates :title, :description, :image_url, presence: true
   validates :title, :permalink, uniqueness: true
 
   validates :image_url, allow_blank: true, url: true
   validates :permalink, format: { with: /\A[-\w]+\z/, message: 'contains special character.' }
-  validates :permalink, format: { with: /[\w]+-[\w]+-[\w]+/, message: 'has wrong format.' }
+
+  #FIXME_AB: user validates length for 3 hyphen separated words
+  validates :permalink_words, length: { minimum: 3 }
```
Comment (bansalakhil) [inline app/models/product.rb:17]: @akshay1y2 overwite the message.

File: app/models/product.rb [hunk 5]
```diff
@@ -25,11 +31,7 @@ def set_discount_price
       self.discount_price = price unless discount_price?
     end
 
-    # ensure that there are no line items referencing this product
-    def ensure_not_referenced_by_any_line_item
-      unless line_items.empty?
-        errors.add(:base, 'Line Items present')
-        throw :abort
-      end
+    def permalink_words
```
Comment (bansalakhil) [inline app/models/product.rb:34]: @akshay1y2 private

File: app/models/user.rb [hunk 6]
```diff
@@ -1,8 +1,12 @@
 class User < ApplicationRecord
+  #FIXME_AB: User should not be destroy if he has any order
+  has_many :orders, dependent: :restrict_with_error
+  has_many :line_items, through: :orders
+
   ADMIN_EMAIL = 'admin@depot.com'
 
   validates :name, presence: true, uniqueness: true
-  validates :email, uniqueness: true, format: { with: URI::MailTo::EMAIL_REGEXP } 
+  validates :email, uniqueness: true, format: { with: URI::MailTo::EMAIL_REGEXP }
```
Comment (bansalakhil) [inline app/models/user.rb:9]: @akshay1y2 whenever you do uniqueness, ensure to consider case insensitivity

**Pull request merged**